saleor/webhook/circuit_breaker/breaker_board.py

```diff
@@ -0,0 +1,221 @@
+
+    def _open_breaker(self, app_id: int) -> str:
+        self.storage.update_open(app_id, int(time.time()), CircuitBreakerState.OPEN)
+        logger.info(
+            "[App ID: %r] Circuit breaker tripped, cooldown is %r [seconds].",
+            app_id,
+            self.cooldown_seconds,
+        )
+        return CircuitBreakerState.OPEN
+
+    def _half_open_breaker(self, app_id: int) -> str:
+        self.storage.update_open(
+            app_id, int(time.time()), CircuitBreakerState.HALF_OPEN
+        )
+        logger.info(
+            "[App ID: %r] Circuit breaker state changed to %s.",
+            app_id,
+            CircuitBreakerState.HALF_OPEN,
+        )
+        return CircuitBreakerState.HALF_OPEN
+
+    def _close_breaker(self, app_id: int) -> str:
+        self.storage.update_open(app_id, 0, CircuitBreakerState.CLOSED)
+        logger.info(
+            "[App ID: %r] Circuit breaker fully recovered.",
+            app_id,
+        )
+        return CircuitBreakerState.CLOSED
+
+    def update_breaker_state(self, app_id: int) -> str:
+        last_open, state = self.storage.last_open(app_id)
+        total = self.storage.get_event_count(app_id, "total") or 1
+        errors = self.storage.get_event_count(app_id, "error")
+        # CLOSED to OPEN
+        if state == CircuitBreakerState.CLOSED and self.exceeded_error_threshold(
+            state, total, errors
+        ):
+            return self.set_breaker_state(app_id, CircuitBreakerState.OPEN)
+        # OPEN TO HALF-OPEN
+        if state == CircuitBreakerState.OPEN and last_open < (
+            time.time() - self.cooldown_seconds
+        ):
+            return self.set_breaker_state(app_id, CircuitBreakerState.HALF_OPEN)
+        # HALF-OPEN to CLOSED / OPEN
+        if state == CircuitBreakerState.HALF_OPEN:
+            if self.exceeded_error_threshold(state, total, errors):
+                return self.set_breaker_state(app_id, CircuitBreakerState.OPEN)
+            if self.reached_half_open_target_success_count(total, errors):
+                return self.set_breaker_state(app_id, CircuitBreakerState.CLOSED)
+        return state
+
+    def register_error(self, app_id: int, ttl: int):
+        self.storage.register_event(app_id, "error", self.ttl_seconds)
+        self.storage.register_event(app_id, "total", self.ttl_seconds)
+
+    def register_success(self, app_id: int, ttl: int):
+        self.storage.register_event(app_id, "total", self.ttl_seconds)
+
+    def __call__(self, func):
+        def inner(*args, **kwargs):
+            event_type: str = args[0]
+            webhook: Webhook = args[2]
+
+            if event_type not in settings.BREAKER_BOARD_SYNC_EVENTS:
+                # Execute webhook without affecting breaker state
+                return func(*args, **kwargs)
+
+            app_id = webhook.app.id
+            state = self.update_breaker_state(app_id)
+            if state == CircuitBreakerState.OPEN:
```

Comment:
What about HALF_OPEN? Right now you treat this as CLOSED. But for circuit breaker, when state is HALF_OPEN, limited number of requests should be invoked.

Reply:
In current impelentation HALF_OPEN is passing webhooks similar as CLOSED, but in this state we have different thresholds. In means that it will close faster if webhook endpoint is still down.

In first iteration we don't plan to throttle traffic and execute only some percentage of webhooks. However we may consider extending the implementation in next released if it's necessary.

Reply:
Okay, but this `HALF_OPEN` current flow should be included in the enum description right? I mean, we don't tell anywhere that HALF_OPEN, means that we don't throttle the webhooks. So someone could assume, that it needs to call the mutation to "unblock" the webhooks (which are still working without any limitations).

Reply:
It will be extra documented but you're right, I'll explicitly state it in the description.

Reply:
Also, do you plan to have a WebhookAsync event, to let know that the specific app was blocked by CircutBreaker?

Reply:
@korycins eventual WebhookAsync event implementation wil be handled as separate ticket
https://linear.app/saleor/issue/SCL-727/add-async-webhook-event-for-app-being-blocked-by-circuit-breaker